### PR TITLE
Rebasing tpu branch on a more recent fairseq upstream commit

### PR DESCRIPTION
**Pull request by taylanbil**
Re-do of https://github.com/pytorch-tpu/fairseq/pull/18/

https://github.com/taylanbil/fairseq/pull/1 is showing equivalency between the 2 branches `tpu-rebase-master` and `taylanbil-tpu-rebase-master`. For some reson, this pr can be merged and the other can't. I cannot git, sorry.

--------

This rebase will enable working on RoBERTa.

Transformer on WMT18 performance is about the same. Convergence is verified via bleu score being 25.83 in epoch 5 (will not merge before verifying e2e convergence) and validation loss going down similarly to current tpu branch

Getting this out there for review as it's a big pr. My commit is the last 2, and second to last one is a giant commit. it's many commits squashed to 1.

### CODE REVIEW

**Comment by taylanbil** [comment]
For completeness, I'm doing an e2e run atm. Will check back on Monday.

**Comment by dlibenzi** [comment]
One flyby comment before review 😉 
Next time, please, one PR for the rebase, and one for your changes.
Also, check your commit comments, and maybe squash them.

**Comment by taylanbil** [comment]
7a23b93 is the squashed version of all my commits over the last months :'( I didn't really intend to do any changes other than resolving conflicts, but it all got messed up in the end, sorry.

File: fairseq/checkpoint_utils.py [hunk 1]
```diff
@@ -62,15 +64,17 @@ def is_better(a, b):
...
 
     checkpoints = [os.path.join(args.save_dir, fn) for fn, cond in checkpoint_conds.items() if cond]
+
     if len(checkpoints) > 0:
         trainer.save_checkpoint(checkpoints[0], extra_state)
         for cp in checkpoints[1:]:
             try:
                 from fairseq.fb_pathmgr import fb_pathmgr
-                fb_pathmgr.copy(checkpoints[0], cp, True)
+                if getattr(args, 'use_gpu', True) or xm.is_master_ordinal():
```
Comment (dlibenzi) [inline fairseq/checkpoint_utils.py]: Maybe move this out of the `try` into a local and reuse.

File: fairseq/checkpoint_utils.py [hunk 2]
```diff
@@ -210,7 +214,8 @@ def checkpoint_paths(path, pattern=r'checkpoint(\d+)\.pt'):
 def torch_persistent_save(*args, **kwargs):
     for i in range(3):
         try:
-            return torch.save(*args, **kwargs)
+            save_func = xm.save if kwargs.pop('xla', False) else torch.save
```
Comment (dlibenzi) [inline fairseq/checkpoint_utils.py]: I am not sure if kwargs here is passed by reference or value, but if the former, you will be changing the object at call site.

File: fairseq/criterions/label_smoothed_cross_entropy.py [hunk 3]
```diff
@@ -17,8 +17,8 @@ def label_smoothed_nll_loss(lprobs, target, epsilon, ignore_index=None, reduce=T
     smooth_loss = -lprobs.sum(dim=-1, keepdim=True)
     if ignore_index is not None:
         non_pad_mask = target.ne(ignore_index)
-        nll_loss = nll_loss[non_pad_mask]
-        smooth_loss = smooth_loss[non_pad_mask]
+        nll_loss.masked_fill_(~non_pad_mask, 0.0)
```
Comment (dlibenzi) [inline fairseq/criterions/label_smoothed_cross_entropy.py]: Need to be careful here, since if the code does an average on the number of tensor elements, a padded one will skew the loss final value.
Was this done to avoid `masked_select()`?

File: fairseq/data/data_utils.py [hunk 4]
```diff
@@ -227,10 +243,25 @@ def batch_by_size(
...
     if isinstance(indices, types.GeneratorType):
         indices = np.fromiter(indices, dtype=np.int64, count=-1)
-
     return batch_by_size_fast(indices, num_tokens_fn, max_tokens, max_sentences, bsz_mult)
 
 
+def batch_by_size_tpu(
+    indices, num_tokens_fn, input_shapes
+):
+    batches = [[] for _ in input_shapes]
```
Comment (dlibenzi) [inline fairseq/data/data_utils.py:274]:
```Python
[[]] * len(...)
```
?

File: fairseq/modules/multihead_attention.py [hunk 5]
```diff
@@ -242,10 +247,9 @@ def forward(
         if key_padding_mask is not None:
             # don't attend to padding symbols
             attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
-            attn_weights = attn_weights.masked_fill(
-                key_padding_mask.unsqueeze(1).unsqueeze(2),
-                float('-inf'),
-            )
+            attn_weights = attn_weights.transpose(0, 2)
```
Comment (dlibenzi) [inline fairseq/modules/multihead_attention.py:257]: Careful here.
Same view issue we chatted on Slack.
A `masked_fill()` and a `masked_fill_()` have different semantics when applied to views and transposes.
View and transposes (as well as a full set of other pytorch ops) have view-semantics.
Which means that in-place ops (the ones ending with underscore) will modify the source of the view.

File: fairseq/models/transformer.py [hunk 6]
```diff
@@ -355,8 +355,8 @@ def forward(self, src_tokens, src_lengths, cls_input=None, return_all_hiddens=Fa
 
         # compute padding mask
         encoder_padding_mask = src_tokens.eq(self.padding_idx)
-        if not encoder_padding_mask.any():
-            encoder_padding_mask = None
+        #if not encoder_padding_mask.any():
```
Comment (dlibenzi) [inline fairseq/models/transformer.py]: Do you need to keep this?

File: fairseq/modules/multihead_attention.py [hunk 7]
```diff
@@ -57,11 +65,12 @@ def __init__(self, embed_dim, num_heads, kdim=None, vdim=None, dropout=0., bias=
 
         self.onnx_trace = False
 
+        # XXX: (taylanbil) try F.multi...
         self.enable_torch_version = False
-        if hasattr(F, "multi_head_attention_forward"):
-            self.enable_torch_version = True
-        else:
-            self.enable_torch_version = False
+        # if hasattr(F, "multi_head_attention_forward"):
```
Comment (dlibenzi) [inline fairseq/modules/multihead_attention.py]: Maybe clean up?

File: fairseq/trainer.py [hunk 8]
```diff
@@ -576,6 +585,14 @@ def get_meter(self, name):
             return None
         return self.meters[name]
 
+    def meters_to_device(self, device):
```
Comment (dlibenzi) [inline fairseq/trainer.py:588]: How many meters there are?
This is sending one tensor at a time, which can be slow.
If there are not many meters, and this gets called only at app init, it might be OK.

File: train.py [hunk 9]
```diff
@@ -301,11 +320,248 @@ def distributed_main(i, args, start_rank=0):
...
+            xm.master_print(met.metrics_report())
+
+    train_meter.stop()
+    xm.master_print('| done training in {:.1f} seconds'.format(train_meter.sum))
+    assert_on_losses(args, trainer)
+
+
+def assert_on_losses(args, trainer):
+    if xu.getenv_as('XLA_USE_BF16', bool, False):
+        # XXX: loss values are meaningless in this case due to precision in bf16
```
Comment (dlibenzi) [inline train.py]: What's this `XXX` about?

**Review by dlibenzi — CHANGES_REQUESTED** [review]
It is pretty hard to review things like this.

**Review by dlibenzi — COMMENTED** [review]
Are these your changes? Did they come with a rebase? Hard to tell 😉

**Review by dlibenzi — CHANGES_REQUESTED** [review]
You may want to squash these commits and fix the comments (not just list the full original comments).

File: fairseq/data/data_utils.py (hunk 4 above)
Comment (taylanbil) [reply fairseq/data/data_utils.py:274]: I don't think that works because

```
>>> x=[[]]*10
>>> x
[[], [], [], [], [], [], [], [], [], []]
>>> x[0].append(5)
>>> x
[[5], [5], [5], [5], [5], [5], [5], [5], [5], [5]]
```

File: fairseq/modules/multihead_attention.py (hunk 5 above)
Comment (taylanbil) [reply fairseq/modules/multihead_attention.py:257]: right, so everything works fine this way.

File: fairseq/models/transformer.py (hunk 6 above)
Comment (taylanbil) [reply fairseq/models/transformer.py]: no, will clean.

File: fairseq/modules/multihead_attention.py (hunk 7 above)
Comment (taylanbil) [reply fairseq/modules/multihead_attention.py]: yes, will do

File: fairseq/trainer.py (hunk 8 above)
Comment (taylanbil) [reply fairseq/trainer.py:588]: there are about 10 or so. Not that big of a deal imho, this function is only called initially once.

File: train.py (hunk 9 above)
Comment (taylanbil) [reply train.py]: this is because if bf16 is enabled, loss values will accumulate, and eventually numerator will grow large, and we'll end up with the `4096 + 10 = 4096` issue in the bf16 world.

I believe you have a pr that allows mixed precision, so this issue should be fixable. Haven't gotten around to using that yet. I may ping you when I get to that, to see how I can proceed.

File: fairseq/criterions/label_smoothed_cross_entropy.py (hunk 3 above)
Comment (taylanbil) [reply fairseq/criterions/label_smoothed_cross_entropy.py]: Yeah, this was one of the first things we did when starting to work on the model. original upstream code masked selects non_pad_mask.

in practice, this function is always called with `reduce = True`, so there's no averaging happening for us. Although I may need to change it to put back masked select when reduce is False for other use cases.

File: fairseq/checkpoint_utils.py (hunk 2 above)
Comment (taylanbil) [reply fairseq/checkpoint_utils.py]: I don't think that's true;

```
>>> def f(**kwargs):
...  return kwargs.pop('a')
...
>>> d = {'a':10, 'b':3}
>>> print(f(**d))
10
>>> d
{'a': 10, 'b': 3}
```

File: fairseq/checkpoint_utils.py (hunk 1 above)
Comment (taylanbil) [reply fairseq/checkpoint_utils.py]: will do

**Review by taylanbil — COMMENTED** [review]
Apologies, I honestly don't know what the proper github etiquette is for prs like this. This is basically months of work upstream, merged with months of work in our fork.

I don't really have any new functionalities, changes etc in this pr. Everything I have here is already reviewed stuff, + merge conflict resolutions. Those resolutions are really the only interesting bits of the pr, but there's no way to isolate them afaik.

I don't think I can squash the upstream commits, can I? I can squash the last 6 commits, but I tried that and failed miserably :'( (#18)

File: fairseq/criterions/label_smoothed_cross_entropy.py (hunk 3 above)
Comment (dlibenzi) [reply fairseq/criterions/label_smoothed_cross_entropy.py]: In general, go back and re-check all the places where you replaced `foo()` with `foo_()`, especially where the foo's inputs have *view semantics*.

File: fairseq/data/data_utils.py (hunk 4 above)
Comment (dlibenzi) [reply fairseq/data/data_utils.py:274]: Oh right!

File: fairseq/trainer.py (hunk 8 above)
Comment (dlibenzi) [reply fairseq/trainer.py:588]: OK, there will be a new API in https://github.com/pytorch/xla/pull/1377 which does batching.

File: train.py (hunk 9 above)
Comment (dlibenzi) [reply train.py]: No, I mean, why `XXX` marking?

**Review by dlibenzi — COMMENTED** [review]
I am not the PITA guy asking to split PRs for no reason.
But in this case, the vast majority of this did not need a review, as it was stuff coming off a rebase.
I think you created this PR from the wrong tip of the repo, which ended up including all the commit IDs from the rebase.

In general, about the code, we need to keep in mind that this will be the code folks will look to understand what we are doing to optimize pytorch code for TPU, so clean code and comments would really help.

You can squash your commits:
```git rebase -i HEAD~8```
Read the documentation about it.
Then force-push.

File: fairseq/data/data_utils.py [hunk 10]
```diff
@@ -227,10 +243,25 @@ def batch_by_size(
...
+    for idx in indices:
+        sample_len = num_tokens_fn(idx)
+        for j, (batch_size, padlen) in enumerate(input_shapes):
```
Comment (jysohn23) [inline fairseq/data/data_utils.py]: Is this assuming that the `input_shapes` list will be sorted by shortest to longest `padlen`?

File: fairseq/data/language_pair_dataset.py [hunk 11]
```diff
@@ -178,6 +179,7 @@ def __init__(
         if self.align_dataset is not None:
             assert self.tgt_sizes is not None, "Both source and target needed when alignments are provided"
         self.append_bos = append_bos
+        self.input_shapes = input_shapes
```
Comment (jysohn23) [inline fairseq/data/language_pair_dataset.py:182]: Optional: maybe we could add a docstring for this guy and clarify how it should be sorted?

File: fairseq/modules/multihead_attention.py [hunk 12]
```diff
@@ -146,23 +155,19 @@ def forward(
             saved_state = None
 
         if self.self_attention:
-            q = self.q_proj(query)
```
Comment (jysohn23) [inline fairseq/modules/multihead_attention.py]: For my understanding, was this for performance improvements? If so did it help?

File: fairseq/trainer.py [hunk 13]
```diff
@@ -433,8 +439,11 @@ def maybe_no_sync():
             self.meters['wpb'].update(ntokens)
             self.meters['bsz'].update(nsentences)
             self.meters['gnorm'].update(grad_norm)
+            # the comparison below introduces too many .item() calls and slows
+            # down tpu
             self.meters['clip'].update(
-                1. if grad_norm > self.args.clip_norm and self.args.clip_norm > 0 else 0.
+                0.
+                #1. if grad_norm > self.args.clip_norm and self.args.clip_norm > 0 else 0.
```
Comment (jysohn23) [inline fairseq/trainer.py:446]: We don't need this comment right and all other comments in the PR?

File: train.py [hunk 14]
```diff
@@ -9,19 +9,56 @@
...
 import random
+from datetime import datetime
 
 import numpy as np
 import torch
-
-from fairseq import checkpoint_utils, distributed_utils, options, progress_bar, tasks, utils
+import torch_xla
+import torch_xla.debug.metrics as met
+import torch_xla.distributed.data_parallel as dp
```
Comment (jysohn23) [inline train.py]: Can we remove unused imports like `data_parallel`?

**Review by jysohn23 — CHANGES_REQUESTED** [review]

File: fairseq/data/data_utils.py (hunk 10 above)
Comment (taylanbil) [reply fairseq/data/data_utils.py]: yes

File: fairseq/data/language_pair_dataset.py (hunk 11 above)
Comment (taylanbil) [reply fairseq/data/language_pair_dataset.py:182]: see https://github.com/pytorch-tpu/fairseq/blob/tpu/train.py#L291-L298, we error while parsing the input args if the shapes passed in doesn't satisfy the assumption, and describe requirements.

File: fairseq/modules/multihead_attention.py (hunk 12 above)
Comment (taylanbil) [reply fairseq/modules/multihead_attention.py]: yes, this was causing a 10% regression.

File: fairseq/trainer.py (hunk 13 above)
Comment (taylanbil) [reply fairseq/trainer.py:446]: yeah, this line causes a _local_scalar_dense, so I replaced it with 0. The meter is essentially invalid. I'll cancel this meter if xla.

File: train.py (hunk 14 above)
Comment (taylanbil) [reply train.py]: yep, will do, thanks.

File: train.py (hunk 9 above)
Comment (taylanbil) [reply train.py]: I'll delete

File: fairseq/criterions/label_smoothed_cross_entropy.py (hunk 3 above)
Comment (taylanbil) [reply fairseq/criterions/label_smoothed_cross_entropy.py]: There in only that one place I already reverted, we should be good now.

File: fairseq/trainer.py (hunk 13 above)
Comment (taylanbil) [reply fairseq/trainer.py:446]: actually, may I handle this in a separate pr? I have one in the works that will change the logic around meters a bit.

**Comment by taylanbil** [comment]
Synced offline with @dlibenzi, I am now adding comments in code in the places I edited vs upstream/master.

File: fairseq/criterions/label_smoothed_cross_entropy.py (hunk 3 above)
Comment (taylanbil) [reply fairseq/criterions/label_smoothed_cross_entropy.py]: double checked this.

**Comment by taylanbil** [comment]
[Here](https://gist.github.com/taylanbil/150abd31b1fbf5c91ca90ef5a4d79f08) is all my changes. I'm in the process of adding comments for them.

**Review by dlibenzi — APPROVED** [review]
Approving, but as I mentioned in another comment, the code here will be used by folks wanting to embark similar models, as reference.
So I wish we could get this a bit more documented.

**Review by jysohn23 — APPROVED** [review]
